pkg/vm/memory/relocatable.go

```diff
@@ -6,89 +6,162 @@ import (
+	if !mv.isAddress {
+		return nil, fmt.Errorf("error trying to read a memory address as field element")
 	}
+	return &MemoryAddress{
+		SegmentIndex: mv.segmentIndex,
+		Offset:       mv.value,
+	}, nil
+}
 
-	if r1.SegmentIndex != r2.SegmentIndex {
-		return nil, fmt.Errorf("cannot subtract relocatables from different segments: %d != %d", r1.SegmentIndex, r2.SegmentIndex)
+func (mv *MemoryValue) ToFieldElement() (*f.Element, error) {
+	if mv.isAddress {
+		return nil, fmt.Errorf("error trying to read a field element as a memory address")
 	}
+	return mv.value, nil
+}
 
-	r.SegmentIndex = feltSegment
-	r.Offset.Sub(&r1.Offset, &r2.Offset)
+func MemoryValueFromMemoryAddress(address *MemoryAddress) *MemoryValue {
+	return &MemoryValue{
+		segmentIndex: address.SegmentIndex,
+		value:        address.Offset,
+		isAddress:    true,
+	}
+}
 
-	return r, nil
+func MemoryValueFromFieldElement(felt *f.Element) *MemoryValue {
+	return &MemoryValue{
+		value:     felt,
+		isAddress: false,
+	}
 }
 
-// String representation of a relocatable.
-func (r MemoryAddress) String() string {
-	return fmt.Sprintf("%d:%d", r.SegmentIndex, r.Offset)
+func MemoryValueFromUint64(v uint64) *MemoryValue {
+	newElement := f.NewElement(v)
+	return &MemoryValue{
+		value: &newElement,
+	}
 }
 
-// Given a map of segment relocation, update a memory address location
-func (r *MemoryAddress) Relocate(r1 *MemoryAddress, segmentsOffsets *map[uint64]*MemoryAddress) (*MemoryAddress, error) {
-	if (*segmentsOffsets)[r1.SegmentIndex] == nil {
-		return nil, fmt.Errorf("missing segment %d relocation rule", r.SegmentIndex)
+// Adds two memory value is the second one is a Felt
+func (memVal *MemoryValue) Add(lhs *MemoryValue, rhs *MemoryValue) (*MemoryValue, error) {
+	if rhs.isAddress {
+		return nil, fmt.Errorf("cannot add two memory addresses")
 	}
 
-	r, err := r.Add((*segmentsOffsets)[r1.SegmentIndex], &MemoryAddress{0, r1.Offset})
+	fmt.Println(memVal)
+	fmt.Println("x0")
+	fmt.Println(lhs)
+	fmt.Println("x1")
+	fmt.Println(rhs)
+
+	// todo(rodro): is lhs always a memory address?
+	if lhs.isAddress {
+		memVal.isAddress = true
+		memVal.segmentIndex = lhs.segmentIndex
+	}
+	fmt.Println("x2", lhs.value, rhs.value)
+	memVal.value.Add(lhs.value, rhs.value)
+	fmt.Println("x3")
 
-	return r, err
+	return memVal, nil
 }
 
-// Turns a relocatable of the form 0:offset into offset.Uint64()
-// otherwise fails.
-func (r *MemoryAddress) Uint64() (uint64, error) {
-	if r.SegmentIndex == feltSegment {
-		return r.Offset.Uint64(), nil
+// Subs two relocatables if they're in the same segment or the rhs is a Felt.
+func (memVal *MemoryValue) Sub(lhs *MemoryValue, rhs *MemoryValue) (*MemoryValue, error) {
+	if !rhs.isAddress {
+		// todo(rodro): is lhs always a memory address?
+		if lhs.isAddress {
+			memVal.isAddress = true
+			memVal.segmentIndex = lhs.segmentIndex
+		}
+		memVal.value.Sub(lhs.value, rhs.value)
+		return memVal, nil
+	}
+
+	// todo(rodro): can lhs not be a memory address?
```

Comment:
I think all operations should be just lifted felt operations if both lhs and rhs are felts

Reply:
Yes I agree. Although I wonder if we are going to ever add/substract two memory value that hold felts.

Reply:
Same as above